src/server/verification/verificationAPI.js

```diff
@@ -395,35 +388,140 @@ const setup = (app: Router, verifier: VerificationAPI, storage: StorageAPI) => {
       let loginToken = user.loginToken
 
       if (!loginToken) {
-        const secureHash = md5(user.email + conf.secure_key)
-        const web3Response = await fetch(`${conf.web3SiteUrl}/api/wl/user`, {
-          method: 'PUT',
-          headers: {
-            'Content-Type': 'application/json'
-          },
-          body: JSON.stringify({
-            secure_hash: secureHash.toLowerCase(),
-            email: user.email
-          })
-        })
-          .then(res => res.json())
-          .catch(e => {
-            logger.error('Get Web3 Login Response Failed', e)
-          })
+        const w3Data = await W3Helper.getLoginOrWalletToken(user)
 
-        const web3ResponseData = web3Response && web3Response.data
+        if (w3Data && w3Data.login_token) {
+          loginToken = w3Data.login_token
 
-        if (web3ResponseData && web3ResponseData.login_token) {
-          loginToken = web3ResponseData.login_token
+          storage.updateUser({ ...user, loginToken })
         }
       }
 
+      logger.info('loginToken', loginToken)
+
       res.json({
         ok: +Boolean(loginToken),
         loginToken
       })
     })
   )
+
+  /**
+   * @api {get} /verify/bonuses check if there is available bonuses to charge on user's wallet and do it
+   * @apiName Web3 Charge Bonuses
+   * @apiGroup Verification
+   *
+   * @apiSuccess {Number} ok
+   * @ignore
+   */
+  app.get(
+    '/verify/w3/bonuses',
+    passport.authenticate('jwt', { session: false }),
+    wrapAsync(async (req, res, next) => {
+      const log = req.log.child({ from: 'verificationAPI - verify/bonuses' })
+
+      const { user: currentUser } = req
+      const isUserWhitelisted = await AdminWallet.isVerified(currentUser.gdAddress)
+
+      if (!isUserWhitelisted) {
+        return res.status(200).json({
+          ok: 0,
+          message: 'User should be verified to get bonuses'
+        })
+      }
+
+      let wallet_token = currentUser.w3Token
+
+      if (!wallet_token) {
+        const w3Data = await W3Helper.getLoginOrWalletToken(currentUser)
+
+        if (w3Data && w3Data.wallet_token) {
+          wallet_token = w3Data.wallet_token
+
+          storage.updateUser({ identifier: currentUser.loggedInAs, w3Token: w3Data.wallet_token })
+        }
+      }
+
+      if (!wallet_token) {
```

Comment:
duplicate condition, look 435 line

Reply:
@StanislavShevchenko it is not a mistake. First time (in 435 line) I am checking if there is a w3Token in user record. If not, then trying to get it from W3 Site. If there's still not w3Token then the server responded with error to the Dapp.